### PR TITLE
Add tests and documentation for datastax driver SSL integration

### LINKED ISSUE #2: Configure SSLOptions and JKS Via Config
This is sort of a feature request/todo.  We needed to add JKS support and wanted to do it via config so we did the below.  Can make PR if have time.

### Config
``` java
import io.micronaut.context.annotation.EachProperty;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;

@EachProperty(value = "cassandra", primary = "default")
@Getter
@Setter
public class CassandraConfigurationProperties {

    @NonNull
    private String[] contactPoints = new String[]{"127.0.0.1"};
    private int port = 9042;
    @NonNull
    private String clusterName = "local-cluster";
    @NonNull
    private String keyspaceName = "keyspace";
    @NonNull
    private String userName = "admin";
    @NonNull
    private String password = "admin";
    private boolean sslEnabled = false;
    private String keystoreLocation;
    private String keystorePassword;
    private String keyPassword;
    private String truststoreLocation;
    private String truststorePassword;
    private String consistencyLevel = "QUORUM";
}
```

### Beans
```java
import com.datastax.driver.core.Cluster;
import com.datastax.driver.core.RemoteEndpointAwareJdkSSLOptions;
import com.datastax.driver.core.SSLOptions;
import com.datastax.driver.core.Session;
import com.datastax.driver.mapping.MappingManager;
import CassandraConfigurationProperties;
import io.micronaut.context.annotation.Bean;
import io.micronaut.context.annotation.Factory;
import io.micronaut.context.event.BeanCreatedEvent;
import io.micronaut.context.event.BeanCreatedEventListener;
import io.micronaut.core.util.StringUtils;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManagerFactory;
import java.io.FileInputStream;
import java.io.IOException;
import java.security.KeyManagementException;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.security.UnrecoverableKeyException;
import java.security.cert.CertificateException;
import java.util.Locale;
import java.util.Optional;

@Factory
@AllArgsConstructor
@Slf4j
public class CassandraBeanFactory implements BeanCreatedEventListener<Cluster.Builder> {

    private static final String JKS = "JKS";
    private static final String TLS = "TLS";
    private final CassandraConfigurationProperties cassandraConfigurationProperties;

    @Bean
    public Session session(final Cluster cluster) {
        final Session session = cluster.connect();
        session.execute(String.format(Locale.ROOT, "USE %s", cassandraConfigurationProperties.getKeyspaceName()));
        return session;
    }

    @Bean
    public MappingManager mappingManager(final Session session) {
        return new MappingManager(session);
    }

    /**
     * Datastax uses codahale metrics, which is the old metrics library.  Needs to support micrometer then we can re-enable.
     *
     * @param event The bean to configure
     * @return the same bean, but with more sprinkles
     */
    @Override
    public Cluster.Builder onCreated(final BeanCreatedEvent<Cluster.Builder> event) {
        Cluster.Builder bean = event.getBean();
        if (cassandraConfigurationProperties.isSslEnabled()) {
            bean.withSSL(getSSLOptions().orElse(RemoteEndpointAwareJdkSSLOptions.builder().build()));
        }
        if (StringUtils.isNotEmpty(cassandraConfigurationProperties.getPassword())
                && StringUtils.isNotEmpty(cassandraConfigurationProperties.getUserName())) {
            bean.withCredentials(cassandraConfigurationProperties.getUserName(), cassandraConfigurationProperties.getPassword());
        }
        return bean
                .withoutJMXReporting()
                .withoutMetrics();
    }

    private Optional<SSLOptions> getSSLOptions() {
        if (cassandraConfigurationProperties.isSslEnabled() &&
                StringUtils.isNotEmpty(cassandraConfigurationProperties.getKeystoreLocation()) &&
                StringUtils.isNotEmpty(cassandraConfigurationProperties.getKeystorePassword()) &&
                StringUtils.isNotEmpty(cassandraConfigurationProperties.getKeyPassword()) &&
                StringUtils.isNotEmpty(cassandraConfigurationProperties.getTruststoreLocation()) &&
                StringUtils.isNotEmpty(cassandraConfigurationProperties.getTruststorePassword())) {
            try {

                KeyStore keyStore = KeyStore.getInstance(JKS);
                keyStore.load(new FileInputStream(cassandraConfigurationProperties.getKeystoreLocation()), cassandraConfigurationProperties.getKeystorePassword().toCharArray());
                KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
                keyManagerFactory.init(keyStore, cassandraConfigurationProperties.getKeyPassword().toCharArray());

                KeyStore trustStore = KeyStore.getInstance(JKS);
                trustStore.load(new FileInputStream(cassandraConfigurationProperties.getTruststoreLocation()), cassandraConfigurationProperties.getTruststorePassword().toCharArray());
                TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
                trustManagerFactory.init(trustStore);

                SSLContext sslContext = SSLContext.getInstance(TLS);
                sslContext.init(keyManagerFactory.getKeyManagers(), trustManagerFactory.getTrustManagers(), new SecureRandom());

//                return new RemoteEndpointAwareNettySSLOptions(sslContext);

                return Optional.of(RemoteEndpointAwareJdkSSLOptions.builder()
                        .withSSLContext(sslContext)
                        .build());


            } catch (NoSuchAlgorithmException | KeyStoreException | IOException | CertificateException | UnrecoverableKeyException | KeyManagementException e) {
                log.error("Unable to initialize SSL for Cassandra connection.", e);
                throw e;
            }
        } else {
            log.warn("Cassandra SSL not enabled or keystore/truststore properties are missing.");
        }

        return Optional.empty();
    }
}

```

### PR DESCRIPTION
**Pull request by wetted**
closes #2 

- [x]  add tests
- [x]  add documentation

[Datastax Java Driver 4.17 docs for SSL integration](https://docs.datastax.com/en/developer/java-driver/4.17/manual/core/ssl/#driver-configuration)
[Datastax Java Driver 4.17 Configuration reference](https://docs.datastax.com/en/developer/java-driver/4.17/manual/core/configuration/reference/)

### CODE REVIEW

File: tests/cassandra-ssl/build.gradle [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    testImplementation("io.micronaut:micronaut-http-client")
+    testImplementation("io.micronaut:micronaut-jackson-databind")
+    testImplementation("io.micronaut:micronaut-management")
+    testImplementation("io.micronaut.reactor:micronaut-reactor")
+
+    testImplementation('org.bouncycastle:bcpkix-jdk18on:1.77')
+    testImplementation("org.testcontainers:spock")
+    testImplementation('org.testcontainers:cassandra')
+
+    testRuntimeOnly("ch.qos.logback:logback-classic")
```
Comment (timyates) [inline tests/cassandra-ssl/build.gradle:27]: These should use the catalogue `libs....` or `mn....`

File: tests/cassandra-ssl/src/test/groovy/example/CassandraSSLConfigSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+@MicronautTest
+class CassandraSSLConfigSpec extends Specification {
+
+    static final Path keyStorePath = Files.createTempFile("cassandra-test-key-store", "pkcs12")
+    static final Path trustStorePath = Files.createTempFile("cassandra-test-trust-store", "jks")
+
+    @Shared @AutoCleanup CassandraContainer cassandraContainer =
+            (CassandraContainer) (new CassandraContainer(DockerImageName.parse("cassandra:latest")))
+                    .withExposedPorts(9042)
```
Comment (timyates) [inline tests/cassandra-ssl/src/test/groovy/example/CassandraSSLConfigSpec.groovy:43]:
```suggestion
    @Shared
    @AutoCleanup
    CassandraContainer cassandraContainer = new CassandraContainer<>(DockerImageName.parse("cassandra:latest"))
            .withExposedPorts(9042)
```

File: tests/cassandra-ssl/src/test/groovy/example/CassandraSSLConfigSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                "test")
+        HttpClient client = embeddedServer.getApplicationContext().createBean(HttpClient, embeddedServer.getURL())
+
+        def response = client.toBlocking().exchange("/health", CassandraHealthIndicator)
+        CassandraHealthIndicator healthIndicator = embeddedServer.getApplicationContext().getBean(CassandraHealthIndicator)
+        HealthResult result = Mono.from(healthIndicator.result).block()
+
+        then:
+        response.status() == HttpStatus.OK
+        result
```
Comment (timyates) [inline tests/cassandra-ssl/src/test/groovy/example/CassandraSSLConfigSpec.groovy:98]: This isn't required.... if it's null, the next line will show that

**Review by timyates — CHANGES_REQUESTED** [review]
Does this work?  Don't we need to do something to add the certificate to cassandra running in the container?

Can we add a test in here that verifies:

1. The connection is using SSL
2. Queries can be made against cassandra

**Comment by wetted** [comment]
Still trying to get this working. See also: #404

**Comment by wetted** [comment]
The Sonarcloud complaint about test coverage is bogus (wrong).

File: src/main/docs/guide/ssl.adoc [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    advanced:
+      ssl-engine-factory:
+        class: DefaultSslEngineFactory
+        cipher-suites:
+          - TLS_RSA_WITH_AES_128_CBC_SHA
+          - TLS_RSA_WITH_AES_256_CBC_SHA
+        trust-store-path: <path to trust store>
+        trust-store-password: 12345
+        key-store-path: <path to key store>
+        key-store-password: 12345
```
Comment (timyates) [inline src/main/docs/guide/ssl.adoc:18]: This is `truststore`, not `trust-store`

And `keystore`, not `key-store`

We don't need the `cypher-suites` in a minimal example either as far as I know... 🤔 

I would take the minimal example from the ssl test, ie:

```yaml
cassandra:
  default:
    advanced:
      ssl-engine-factory:
                class: DefaultSslEngineFactory
                truststore-path: <path to trust store>
                truststore-password: 12345
```

Comment (timyates) [reply src/main/docs/guide/ssl.adoc:18]: I don't think we need any more of the config that follows

File: src/main/docs/guide/ssl.adoc [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    basic:
+      contact-points:
+        - "127.0.0.1:9042"
+        - "127.0.0.2:8042"
+      load-balancing-policy:
+        local-datacenter: datacenter1
+----
+
+- class: The class of the factory. `DefaultSslEngineFactory` is the default built-in implementation
+- cipher-suites: cipher suites to enable when creating a connection
```
Comment (timyates) [inline src/main/docs/guide/ssl.adoc:31]: I'd remove this

File: src/main/docs/guide/ssl.adoc [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+- trust-store and key-store properties: these are optional and if not present the system property configuration is used instead
```
Comment (timyates) [inline src/main/docs/guide/ssl.adoc:32]: Make sure these are spelled correctly as per the docs.  keystore isn't required either in the tests, so I'd remove it

**Review by timyates — CHANGES_REQUESTED** [review]
Couple of tweaks suggested for the docs?

**Review by timyates — APPROVED** [review]

**Pull request merged**